packages/plugin-hardhat/src/deploy-beacon-proxy.ts

```diff
@@ -0,0 +1,91 @@
+import type { HardhatRuntimeEnvironment } from 'hardhat/types';
+import { ContractFactory, Contract } from 'ethers';
+
+import { Manifest, logWarning, ProxyDeployment } from '@openzeppelin/upgrades-core';
+
+import {
+  DeployOptions,
+  deploy,
+  DeployTransaction,
+  getBeaconProxyFactory,
+  ContractAddressOrInstance,
+  getContractAddress,
+  ContractFactoryOrSigner,
+  getSigner,
+} from './utils';
+import { getInitializerData } from './deploy-proxy';
+import { getImplementationAddressFromBeacon, getInterfaceFromManifest } from './utils/impl-address';
+
+export interface DeployBeaconProxyFunction {
+  (
+    beacon: ContractAddressOrInstance,
+    ImplFactoryOrSigner: ContractFactoryOrSigner,
+    args?: unknown[],
+    opts?: DeployOptions,
+  ): Promise<Contract>;
+  (
+    beacon: ContractAddressOrInstance,
+    ImplFactoryOrSigner: ContractFactoryOrSigner,
+    opts?: DeployOptions,
+  ): Promise<Contract>;
+}
+
+export function makeDeployBeaconProxy(hre: HardhatRuntimeEnvironment): DeployBeaconProxyFunction {
+  return async function deployBeaconProxy(
+    beacon: ContractAddressOrInstance,
+    ImplFactoryOrSigner: ContractFactoryOrSigner,
+    args: unknown[] | DeployOptions = [],
+    opts: DeployOptions = {},
+  ) {
+    if (!Array.isArray(args)) {
+      opts = args;
+      args = [];
+    }
+
+    const { provider } = hre.network;
+    const manifest = await Manifest.forNetwork(provider);
+
+    opts.kind = 'beacon';
+
+    const beaconAddress = getContractAddress(beacon);
+    const implAddress = await getImplementationAddressFromBeacon(hre, getSigner(ImplFactoryOrSigner), beaconAddress);
+    let contractInterface = await getInterfaceFromManifest(hre, implAddress);
+
+    if (contractInterface === undefined) {
+      if (ImplFactoryOrSigner instanceof ContractFactory) {
```

Comment:
Here we use the user provided beacon address to look up the implementation address, then get the interface that we previously stored in the manifest for that implementation address.  Is there any possibility that the manifest is incorrect, aside from corruption?  Considering the implementation itself is immutable.

Although since the interface is used to create the initializer call data, and if both are provided by the user, I agree it makes sense to use both together.  Then if there is a mismatch during deployment, there would be an error anyways.